sys/kern/signal.c

```diff
@@ -300,23 +311,23 @@ void sig_post(signo_t sig) {
   if (sa->sa_handler == SIG_DFL && defact(sig) == SA_STOP) {
     klog("Stopping thread %lu in process PID(%d)", td->td_tid, p->p_pid);
     p->p_state = PS_STOPPED;
-    p->p_flags &= ~PF_CONTINUED;
-    p->p_flags |= PF_STOPPED;
-    cv_broadcast(&p->p_parent->p_waitcv);
-    proc_unlock(p);
-    WITH_MTX_LOCK (&p->p_parent->p_lock)
+    p->p_flags |= PF_STATE_CHANGED;
+    WITH_PROC_LOCK(p->p_parent) {
+      proc_wakeup_parent(p->p_parent);
```

Comment:
Why waking up process is not a part of `sig_kill` responsibility?

Reply:
Because sending a signal doesn't change the state of the process. For instance, I can send `SIGSTOP` and `SIGKILL`, and if `SIGKILL` is delivered first, then the process won't stop at any time.

Reply:
`SIGCONT` is an exception: the act of sending it changes the state of the target process, so we notify the parent in `sig_kill()` in that case.

Reply:
Should not `sig_kill(p, SIGCHLD)` attempt to wake up `p` if `p` is sleeping on `p_waitcv` ?

Reply:
Sending a signal is one thing, and notifying the parent process of a child's state change is another. An ordinary process can send `SIGCHLD` to another process, and that doesn't mean that a child has changed state.

Reply:
You're right! Thanks!